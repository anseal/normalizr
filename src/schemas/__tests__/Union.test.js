// eslint-env jest
import { normalize, schema } from '../../'

describe('UnionSchema normalization', () => {
	test('throws if not given a schemaAttribute', () => {
		expect(() => new schema.Union({})).toThrow()
	})

	test('normalizes an object using string schemaAttribute', () => {
		const user = new schema.Entity('users')
		const group = new schema.Entity('groups')
		const union = new schema.Union(
			{
				users: user,
				groups: group,
			},
			'type'
		)

		expect(normalize({ id: 1, type: 'users' }, union)).toMatchSnapshot()
		expect(normalize({ id: 2, type: 'groups' }, union)).toMatchSnapshot()
	})

	test('normalizes an array of multiple entities using a function to infer the schemaAttribute', () => {
		const user = new schema.Entity('users')
		const group = new schema.Entity('groups')
		const union = new schema.Union(
			{
				users: user,
				groups: group,
			},
			(input) => {
				return input.username ? 'users' : input.groupname ? 'groups' : null
			}
		)

		expect(normalize({ id: 1, username: 'Janey' }, union)).toMatchSnapshot()
		expect(normalize({ id: 2, groupname: 'People' }, union)).toMatchSnapshot()
		expect(normalize({ id: 3, notdefined: 'yep' }, union)).toMatchSnapshot()
	})
})
