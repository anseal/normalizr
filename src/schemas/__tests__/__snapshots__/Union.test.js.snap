// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnionSchema normalization normalizes an array of multiple entities using a function to infer the schemaAttribute 1`] = `
Object {
  "entities": Object {
    "users": Object {
      "1": Object {
        "id": 1,
        "username": "Janey",
      },
    },
  },
  "result": Object {
    "id": 1,
    "schema": "users",
  },
}
`;

exports[`UnionSchema normalization normalizes an array of multiple entities using a function to infer the schemaAttribute 2`] = `
Object {
  "entities": Object {
    "groups": Object {
      "2": Object {
        "groupname": "People",
        "id": 2,
      },
    },
  },
  "result": Object {
    "id": 2,
    "schema": "groups",
  },
}
`;

exports[`UnionSchema normalization normalizes an array of multiple entities using a function to infer the schemaAttribute 3`] = `
Object {
  "entities": Object {},
  "result": Object {
    "id": 3,
    "notdefined": "yep",
  },
}
`;

exports[`UnionSchema normalization normalizes an object using string schemaAttribute 1`] = `
Object {
  "entities": Object {
    "users": Object {
      "1": Object {
        "id": 1,
        "type": "users",
      },
    },
  },
  "result": Object {
    "id": 1,
    "schema": "users",
  },
}
`;

exports[`UnionSchema normalization normalizes an object using string schemaAttribute 2`] = `
Object {
  "entities": Object {
    "groups": Object {
      "2": Object {
        "id": 2,
        "type": "groups",
      },
    },
  },
  "result": Object {
    "id": 2,
    "schema": "groups",
  },
}
`;
